jira_integration_example/tests/jira_notification_handler_integration_test.py

```diff
@@ -0,0 +1,168 @@
+
+    yield metric_descriptor
+
+    # tear down
+    metric_client.delete_metric_descriptor(metric_descriptor.name)
+
+
+@pytest.fixture(scope='function')
+def notification_channel():
+    # setup
+    notification_channel_client = monitoring_v3.NotificationChannelServiceClient()
+    gcp_project_path = notification_channel_client.project_path(constants.PROJECT_ID)
+
+    notification_channel = notification_channel_client.create_notification_channel(
+        gcp_project_path,
+        constants.TEST_NOTIFICATION_CHANNEL)
+    notification_channel = call_get_notification_channel(notification_channel_client, notification_channel.name)
+
+    yield notification_channel
+
+    # tear down
+    notification_channel_client.delete_notification_channel(notification_channel.name)
+
+
+@pytest.fixture(scope='function')
+def alert_policy(metric_descriptor, notification_channel):
+    # setup
+    policy_client = monitoring_v3.AlertPolicyServiceClient()
+    gcp_project_path = policy_client.project_path(constants.PROJECT_ID)
+
+    print(metric_descriptor.name)
+    print(notification_channel.name)
+
+    test_alert_policy = constants.TEST_ALERT_POLICY_TEMPLATE
+    test_alert_policy['notification_channels'].append(notification_channel.name)
+
+    alert_policy = policy_client.create_alert_policy(
+        gcp_project_path,
+        test_alert_policy)
+    alert_policy = call_get_alert_policy(policy_client, alert_policy.name)
+
+    yield alert_policy
+
+    # tear down
+    policy_client.delete_alert_policy(alert_policy.name)
+
+
+def append_to_time_series(point_value):
+    client = monitoring_v3.MetricServiceClient()
+    gcp_project_path = client.project_path(constants.PROJECT_ID)
+
+    series = monitoring_v3.types.TimeSeries()
+    series.metric.type = constants.METRIC_PATH
+    series.resource.type = constants.RESOURCE_TYPE
+    series.resource.labels['instance_id'] = constants.INSTANCE_ID
+    series.resource.labels['zone'] = constants.ZONE
+    point = series.points.add()
+    point.value.double_value = point_value
+    now = time.time()
+    point.interval.end_time.seconds = int(now)
+    point.interval.end_time.nanos = int(
+        (now - point.interval.end_time.seconds) * 10**9)
+
+    client.create_time_series(gcp_project_path, [series])
+
+
+def test_end_to_end(config, metric_descriptor, notification_channel, alert_policy, jira_client):
+    assert metric_descriptor.type == constants.TEST_METRIC_DESCRIPTOR['type']
+    assert notification_channel.display_name == constants.TEST_NOTIFICATION_CHANNEL['display_name']
+    assert alert_policy.display_name == constants.ALERT_POLICY_NAME
+    assert alert_policy.user_labels == constants.TEST_ALERT_POLICY_TEMPLATE['user_labels']
+    assert alert_policy.notification_channels[0] == notification_channel.name
+
+    # trigger incident and check jira issue created
+    append_to_time_series(constants.TRIGGER_NOTIFICATION_THRESHOLD_DOUBLE + 1)
```

Comment:
Will it trigger if you don't add one? E.g. `append_to_time_series(constants.TRIGGER_NOTIFICATION_THRESHOLD_DOUBLE)`

if so, then it's preferable to use edge values, to detect off by one errors (in this case, I guess the off by one error would be in the test code itself)? And, since that tends to be a convention, a reader would probably assume that you added 1 because it's required. So it would be unexpected if the + 1 wasn't required.

Reply:
I think using the edge value won't trigger an incident, since the condition is that the policy is violated when the metric is above the threshold (I'm guessing this is strictly above). Would it be clearer to use a different value to test?